From 44febe5ebf7dfdabc81551344b15ced796e288fb Mon Sep 17 00:00:00 2001
From: Li Wencheng <liwencheng@phytium.com.cn>
Date: Wed, 7 May 2025 17:22:12 +0800
Subject: [PATCH 40/67] PHYTIUM: net/phytmac: Enable the tail pointer by the
 driver

It is need to enable the RX/TX tail pointer to ensure the NIC
works properly.

Signed-off-by: Li Wencheng <liwencheng@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Change-Id: I6e7fbf04484539add0834d0aaa0fee1cb0001355

Link: https://gitee.com/phytium_embedded/phytium-linux-kernel/commit/e1f3cb62ca1bb3fc7267def60f2f853ad27d2280
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/net/ethernet/phytium/phytmac.h    | 2 +-
 drivers/net/ethernet/phytium/phytmac_v2.c | 8 ++++++++
 drivers/net/ethernet/phytium/phytmac_v2.h | 7 +++++++
 3 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/phytium/phytmac.h b/drivers/net/ethernet/phytium/phytmac.h
index 5ea2a383c107..abe43324cd6a 100644
--- a/drivers/net/ethernet/phytium/phytmac.h
+++ b/drivers/net/ethernet/phytium/phytmac.h
@@ -16,7 +16,7 @@

 #define PHYTMAC_DRV_NAME		"phytium-mac"
 #define PHYTMAC_DRV_DESC		"PHYTIUM Ethernet Driver"
-#define PHYTMAC_DRIVER_VERSION		"1.0.44"
+#define PHYTMAC_DRIVER_VERSION		"1.0.45"
 #define PHYTMAC_DEFAULT_MSG_ENABLE	  \
 		(NETIF_MSG_DRV		| \
 		NETIF_MSG_PROBE	| \
diff --git a/drivers/net/ethernet/phytium/phytmac_v2.c b/drivers/net/ethernet/phytium/phytmac_v2.c
index 7b337d13329f..2e9d14dcb9f3 100644
--- a/drivers/net/ethernet/phytium/phytmac_v2.c
+++ b/drivers/net/ethernet/phytium/phytmac_v2.c
@@ -159,8 +159,16 @@ static int phytmac_v2_init_hw(struct phytmac *pdata)
 	u16 cmd_id, cmd_subid;
 	struct phytmac_dma_info dma;
 	struct phytmac_eth_info eth;
+	u32 ptrconfig = 0;
 	u8 mdc;

+	if (pdata->capacities & PHYTMAC_CAPS_TAILPTR)
+		ptrconfig |= PHYTMAC_BIT(TXTAIL_EN);
+	if (pdata->capacities & PHYTMAC_CAPS_RXPTR)
+		ptrconfig |= PHYTMAC_BIT(RXTAIL_EN);
+
+	PHYTMAC_WRITE(pdata, PHYTMAC_TAILPTR_ENABLE, ptrconfig);
+
 	if (pdata->mii_bus) {
 		cmd_id = PHYTMAC_MSG_CMD_SET;
 		cmd_subid = PHYTMAC_MSG_CMD_SET_ENABLE_MDIO;
diff --git a/drivers/net/ethernet/phytium/phytmac_v2.h b/drivers/net/ethernet/phytium/phytmac_v2.h
index 57594732fab9..1303bde7cff0 100644
--- a/drivers/net/ethernet/phytium/phytmac_v2.h
+++ b/drivers/net/ethernet/phytium/phytmac_v2.h
@@ -17,6 +17,7 @@ extern struct phytmac_hw_if phytmac_2p0_hw;
 #define PHYTMAC_RX_MSG_TAIL				0x00c
 #define PHYTMAC_MSG_IMR					0x020
 #define PHYTMAC_MSG_ISR					0x02c
+#define PHYTMAC_TAILPTR_ENABLE				0x038

 #define PHYTMAC_SIZE					0x0048
 #define PHYTMAC_NETWORK_STATUS				0x0240
@@ -43,6 +44,12 @@ extern struct phytmac_hw_if phytmac_2p0_hw;
 #define PHYTMAC_MSG_COMPLETE_INDEX			0
 #define PHYTMAC_MSG_COMPLETE_WIDTH			1

+/* Bitfields in PHYTMAC_TAILPTR_ENABLE */
+#define PHYTMAC_TXTAIL_EN_INDEX		0	/* Enable tx tail */
+#define PHYTMAC_TXTAIL_EN_WIDTH		1
+#define PHYTMAC_RXTAIL_EN_INDEX		16	/* Enable rx tail */
+#define PHYTMAC_RXTAIL_EN_WIDTH		1
+
 /* Bitfields in PHYTMAC_SIZE */
 #define PHYTMAC_MEM_SIZE_INDEX				0
 #define PHYTMAC_MEM_SIZE_WIDTH				4
--
2.50.1
